pycytominer_transform/convert.py

```diff
@@ -0,0 +1,562 @@
+        Grouped records which include metadata about destination filepath
+        where parquet file was written.
+    """
+
+    # gather records to be processed
+    records = gather_records(
+        source_path=source_path,
+        source_datatype=source_datatype,
+        targets=targets,
+        **kwargs,
+    )
+
+    if not isinstance(records, Dict):
+        records = records.result()
+
+    results = {}
+    # for each group of records, map writing parquet per file
+    for record_group_name, record_group in records.items():
+
+        # read files
+        record_group = read_file.map(record=record_group)
+
+        # if the record group has more than one record, we will need a unique name
+        unique_name = False
+        if len(record_group) >= 2:
+            unique_name = True
+
+        # map for writing parquet files with list of files via records
+        destinations = write_parquet.map(
+            record=record_group,
+            dest_path=unmapped(dest_path),
+            unique_name=unmapped(unique_name),
+        )
+
+        if concat:
+            # build a new record group
+            results[record_group_name] = concat_record_group.submit(
+                record_group=destinations,
+                dest_path=dest_path,
+                infer_common_schema=infer_common_schema,
+            )
+        else:
+            results[record_group_name] = destinations
+
+    return {
+        key: value.result()
+        if isinstance(value, PrefectFuture)
+        else [inner_result.result() for inner_result in value]
+        if not isinstance(value, Dict)
+        else value
+        for key, value in results.items()
+    }
+
+
+def convert(  # pylint: disable=too-many-arguments
+    source_path: str,
+    dest_path: str,
+    dest_datatype: Literal["parquet"],
+    source_datatype: Optional[str] = None,
+    targets: Union[List[str], Tuple[str, ...]] = DEFAULT_TARGETS,
+    concat: bool = True,
+    infer_common_schema: bool = True,
+    task_runner: BaseTaskRunner = ConcurrentTaskRunner,
+    **kwargs,
+) -> Dict[str, List[Dict[str, Any]]]:
+    """
+    Convert file-based data from various sources to Pycytominer-compatible standards.
+
+    Note: source paths may be object-storage locations using common "s3://..." or similar.
+
+    Args:
+      source_path: Union[str, pathlib.Path, AnyPath]:
+        str or Path-like reference to read source files from.
+      dest_path: str:
+        Path to write files to.
+      dest_datatype: Literal["parquet"]:
+        Destination datatype to write to.
+      source_datatype: Optional[str]:  (Default value = None)
+        Source datatype to focus on during conversion.
+      targets: Union[List[str], Tuple[str, str, str, str]]:  (Default value = None)
+        Target filenames to use for conversion.
+      concat: bool:  (Default value = True)
+        Whether to concatenate similar files together.
+      infer_common_schema: bool (Default value = True)
```

Comment:
Thank you for this comment, I need to do a better job explaining this. For each data source group, there are two paths for successful concatenation (making two files with similar columns one continuous dataset):

1. Use the "first" file discovered within a group as the basis for all other schema within the group, throwing an error if any mismatch is discovered.
2. Use all files within the group to form a "minimum viable schema" as the basis for concatenation, ignoring columns and data which exist in some but not all files. With this we must defer to using only columns which are in all the files. 

With option 2 my thoughts were that some data when concatenating would be better than none. It could be that raising an exception on any mismatch is more useful. What do you think about this? Please let me know if I may clarify any of the above here.

Reply:
We made a decision in cytominer-database to provide two options. See the notes under "There are two options for the key reference_option" here: https://github.com/cytomining/cytominer-database/blob/9cf43400a01b82e4b4bce0901e039b7f24519ffe/README.rst

We'll probably want `infer_common_schema=True` to be close to the `reference_option = sample` option  and `infer_common_schema=False` to be close to the `reference_option = path/to/reference/folder` option.

I don't think we need to solve both of these options in the present PR, so maybe what you have currently is ok!

Reply:
Thank you for linking to that document, it will be helpful moving forward here. We may have achieved roughly the same for `infer_common_schema=True / reference_option = sample` without randomness (if this is useful). On the other hand, for `infer_common_schema=False /  reference_option = path/to/reference/folder`, providing a specific reference file option would be a great way to avoid surprises! I've added a new issue to help make sure we address and review this with #16 .